Dockerfile

```diff
@@ -0,0 +1,39 @@
+# syntax=docker/dockerfile:1
+
+# Stage 1 - Build the toolchain
+FROM ubuntu:18.04
+ARG N64_INST=/n64_toolchain
+ENV N64_INST=${N64_INST}
+
+# install dependencies
+RUN apt-get update
+RUN apt-get install -yq wget bzip2 gcc g++ make file libmpfr-dev libmpc-dev zlib1g-dev texinfo git gcc-multilib
+
+# Build
+COPY ./tools/build-toolchain.sh /tmp/tools/build-toolchain.sh
+WORKDIR /tmp/tools
+RUN ./build-toolchain.sh
```

Comment:
I have one question here. You're copying the shell file into `/tmp` which I think it defeats Docker caching. Since the script is in tmpfs, these steps will always be executed. Thus, every time you run `docker build` on this Dockerfile, I believe it will rebuild the toolchain from scatch without caching.

If you instead copy the `build-toolchain.sh` script to the filesystem, I *think* it should then use it as part of the caching. That is, I would expect that if the script has not changed, the build will be skipped as cached.

Reply:
Docker does not perform file system comparisons it only looks at the copied file's hashes so his gets properly cached. For the `RUN`s it does only look at the command itself. Also see [here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache)